crates/memory/src/op_store/time_slice_hash_store.rs

```diff
@@ -0,0 +1,140 @@
+use kitsune2_api::{K2Error, K2Result};
+use std::collections::BTreeMap;
+
+/// In-memory store for time slice hashes.
+///
+/// Empty hashes do not need to be stored, and will be rejected with an error if you try to insert
+/// one. Hashes that are stored, are stored sparsely, and are indexed by the slice id.
+///
+/// It is valid to look up a time slice which has not had a hash stored, and you will get a `None`
+/// response. Otherwise, you will get exactly what was most recently stored for that slice id.
+#[derive(Debug, Default)]
+#[cfg_attr(test, derive(Clone))]
+pub(super) struct TimeSliceHashStore {
+    inner: BTreeMap<u64, bytes::Bytes>,
+}
+
+impl TimeSliceHashStore {
+    /// Insert a hash at the given slice id.
+    pub(super) fn insert(
+        &mut self,
+        slice_id: u64,
+        hash: bytes::Bytes,
+    ) -> K2Result<()> {
+        // This doesn't need to be supported. If we receive an empty hash
+        // for a slice id after we've already stored a non-empty hash for
+        // that slice id, then the caller has done something wrong.
+        // Alternatively, if we've computed an empty hash for a time slice, then we don't
+        // need to store that.
+        if hash.is_empty() {
+            return Err(K2Error::other("Cannot insert empty combined hash"));
+        }
+
+        self.inner.insert(slice_id, hash);
+
+        Ok(())
+    }
+
+    pub(super) fn get(&self, slice_id: u64) -> Option<bytes::Bytes> {
+        self.inner.get(&slice_id).cloned()
+    }
+
+    pub fn highest_stored_id(&self) -> Option<u64> {
+        self.inner.iter().last().map(|(id, _)| *id)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn create_empty() {
+        let store = TimeSliceHashStore::default();
+
+        assert_eq!(None, store.highest_stored_id());
+        assert!(store.inner.is_empty());
+    }
+
+    #[test]
+    #[should_panic(expected = "Cannot insert empty combined hash")]
```

Comment:
Seems like this doesn't panic but returns an error.

Reply:
I can write it as a test that errors